caffe2/operators/reciprocal_op.cc

```diff
@@ -0,0 +1,96 @@
+#include "caffe2/operators/reciprocal_op.h"
+
+#include <string>
+#include <vector>
+
+namespace caffe2 {
+
+REGISTER_CPU_OPERATOR(
+    Reciprocal,
+    UnaryElementwiseOp<
+        TensorTypes<float>,
+        CPUContext,
+        ReciprocalFunctor<CPUContext>>);
+
+// Input: X, output: Y
+OPERATOR_SCHEMA(Reciprocal)
+    .NumInputs(1)
+    .NumOutputs(1)
+    .AllowInplace({{0, 0}})
+    .IdenticalTypeAndShape()
+    .SetDoc(R"DOC(
+Performs element-wise reciprocal ($\1/x$) of input tensor $X$.
+
+Github Link:
+- https://github.com/pytorch/pytorch/blob/master/caffe2/operators/reciprocal_op.cc
+
+<details>
+
+<summary> <b>Example</b> </summary>
+
+**Code**
+
+```
+
+workspace.ResetWorkspace()
+
+op = core.CreateOperator(
+    "Reciprocal",
+    ["X"],
+    ["Y"],
+)
+
+workspace.FeedBlob("X", (np.random.randint(10, size=(3,3))).astype(np.float32))
+print("X:", workspace.FetchBlob("X"))
+workspace.RunOperatorOnce(op)
+print("Y:", workspace.FetchBlob("Y"))
+
+```
+
+**Result**
+
+```
+
+X:
+[[8. 3. 3.]
+ [4. 0. 0.]
+ [1. 2. 5.]]
+Y:
+[[0.125 0.3333333  0.3333333 ]
+ [0.25  inf        inf       ]
+ [1     0.5        0.2       ]]
+
+```
+
+</details>
+)DOC")
+.Input(0, "X", "*(type: Tensor`<float>`)* Input data tensor.")
+.Output(0, "Y", "*(type: Tensor`<float>`)* Output tensor.");
+
+namespace {
+
+class GetReciprocalGradient : public GradientMakerBase {
+  using GradientMakerBase::GradientMakerBase;
+  std::vector<OperatorDef> GetGradientDefs() override {
+    Argument scale_arg;
+    scale_arg.set_name("scale");
+    scale_arg.set_f(0.5);
+    return std::vector<OperatorDef>{CreateOperatorDef(
```

Comment:
I'm a little confused here. For Y = 1 / X. I think dY/dX = -1/(X^2) = -Y^2. But why here you are using Div and Scale op to simulate the gradient?

Reply:
I'm not sure I understand what do you mean here. I'm using the same approach as in sqrt operator. What these div/scale mean and how it correlates to dy/dx?

Reply:
Maybe you didn't get the design of Ops here. Let me explain it a little more.
In most cases, we need both the XXOp and XXGradientOp for forward and backward path. For the backward path, we need to compute dL/dX by given dL/dY and X or Y. Here GO(0) means dL/dY, and GI(0) means dL/dX. One example is RsqrtOp https://github.com/pytorch/pytorch/blob/master/caffe2/operators/rsqrt_op.cc

However, in some cases, the backward path can be formulated in some existing ops. For example the SqrtOp, Y = X^0.5. so dL/dX = dL/dY * dY/dX = dL/dY * 0.5 * X^(-0.5) = dL/dY * 0.5 / Y; That's why they use a ScaleOp and another Div Op to simulate it. So here you have to implement by yourself. 

I'd suggest write it in another GradientOp instead of using ScaleOp and SquareOp to simulate it. Since for GPU side, that will lead to launch the kernel twice while if we write it by ourselves, it can launch the kernel only once.